src/preact/base-element.js

```diff
@@ -375,6 +416,13 @@ export class PreactBaseElement extends AMP.BaseElement {
           shadowRoot.appendChild(serviceSlot);
         }
         this.container_ = container;
+
+        // Connect shadow root to the element's context.
+        setParent(shadowRoot, this.element);
+        // TODO(#30283): in Shadow DOM, only the children distributed in
+        // slots are displayed. All other children are undisplayed. We need
```

Comment:
is it expected that we will have undistributed amp elements?

Reply:
It's very possible from the point of view of a generic proto `PreactBaseElement`. It relies on the underlying component to render a slot. If it doesn't - the matching node is undistributed. There's a latency between this call and when a component gets rendered. And also, it might be a component's logic to leave some slots unrendered. For instance, a carousel might decide to only render 5 slides at any given time and placeholders for the remainder. This could improve performance. From DOM APIs the situation is quite good in this case. When slot distribution changes (new nodes get slotted or previous nodes get unslotted) we get and event (see https://github.com/ampproject/amphtml/pull/30321) and we can find the necessary information (what's shown and what's not) from this event. What this comment says is that we will likely want to have one "fast and clean" way to deal with all unslotted nodes as a group, because we know these nodes cannot be shown.

This might seem new, but it's something we do right now anyway. The old carousel code applies `display:none` to the slides too far away from the current slide. This is not much different. And in a way more semantic. We don't know what `display:none` means usually, but unslotted simply means that a node is not used by the shadow root at this time.